### PR TITLE
[ BB2-1119 ] Settings configuration via JSON or YAML files

### PR DESCRIPTION
**Pull request by dtisza1**
**JIRA Ticket:**
[BB2-1119](https://jira.cms.gov/browse/BB2-1119)

**User Story or Bug Summary:**

Python SDK: Settings configuration via JSON or YAML files:

  - Provides configuration of the SDK via configuration files and related tests.

### What Does This PR Do?

**The following is a summary of the files in this PR**

* Create SDK JSON configuration file functionality
  - Add SDK user config files to `.gitignore`
  - Create `Bb2` class methods for reading settings from JSON configuration files in `src/bb2.py`.
  - Create `src/tests/test_configs.py` with configuration and JSON related file tests.
  - Create JSON test files to be used in tests under `src/tests/test_configs/json/`.
  - Create JSON sample configuration for SDK user in `bluebutton-sample-config.json`.

* Add YAML format configuration file functionality
  - Add `.yaml` file extension support to `Bb2` class methods in `src/bb2.py`.
  - Add tests for YAML configuration files in `src/tests/test_configs.py`.
  - Create YAML test files to be used in tests under `src/tests/test_configs/yaml/`.
  - Create YAML sample configuration for SDK user in `bluebutton-sample-config.yaml`.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check these things, in particular:

* Are there any unhandled and/or untested edge cases you can think of?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
* Naming in general? Do the namings of variables, methods, classes, and other items make sense? Are they too verbose, or not verbose enough? Any redundancies? Are the comments descriptive enough?
* Are audit logging items properly handled? Is sensitive information either not included or hashed?

* *What/how to review:*

### CODE REVIEW

File: src/bb2.py [hunk 1]
```diff
@@ -2,13 +2,90 @@
 Blue Button 2.0 SDK Class
 
 """
+import json
+import os
+import pathlib
+import yaml
+
+ROOT_DIR = os.path.abspath(os.curdir) + "/"
```
Comment (nbragdon) [inline src/bb2.py:10]: Just curious - if I import this package, and then run my python server NOT from the top level... will this still work and find my top level config?

File: src/bb2.py [hunk 2]
```diff
@@ -2,13 +2,90 @@
...
     name = "bb2"
     verbose_name = "Blue Button 2.0 SDK Package"
-    
-    def hello(self):
-        mesg = "Hello from BB2 SDK Class method!!!"
-        print(mesg)
-        return mesg
\ No newline at end of file
+
+    def __init__(self, config_file=DEFAULT_CONFIG_FILE_LOCATION):
```
Comment (nbragdon) [inline src/bb2.py:23]: So this constructor will only support a config file location, not passing in the config as a dict?

File: src/bb2.py [hunk 3]
```diff
@@ -2,13 +2,90 @@
...
+            raise ValueError(
+                "Error: Configuration file extension must be .json"
+                " or .yaml for: {}".format(
+                    self.config_file
+                )
+            )
+
+    def set_configuration(self, config):
+        # Check environment setting
+        env = config.get("environment", None)
```
Comment (nbragdon) [inline src/bb2.py:66]: What do you think of defaulting to sandbox?
```suggestion
        env = config.get("environment", "SANDBOX")
```

File: src/bb2.py [hunk 4]
```diff
@@ -2,13 +2,90 @@
...
-        mesg = "Hello from BB2 SDK Class method!!!"
-        print(mesg)
-        return mesg
\ No newline at end of file
+
+    def __init__(self, config_file=DEFAULT_CONFIG_FILE_LOCATION):
+        self.config_file = config_file
+        self.config = {}
+
+        self.environment = None
```
Comment (nbragdon) [inline src/bb2.py:27]: Do we actually need to keep track of the env? I know for the node SDK I just used it to generate the base url but never actually saved it.

File: bluebutton-sample-config.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (nbragdon) [inline bluebutton-sample-config.json:1]: Just a reminder that these files will be inside the SDK and the 3rd party app wont have access to them (just think when you install a library via pip, you don't generally copy files out of it). So I'm not sure here is where we want our sample files, probably just show how to create them in the read me.

File: src/tests/test_configs/yaml/bluebutton-sample-config-missing-callback.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# BB2 target service SANDBOX or PRODUCTION
```
Comment (nbragdon) [inline src/tests/test_configs/yaml/bluebutton-sample-config-missing-callback.yaml:1]: Because you created generic code, you could actually decrease the number of test files you have here. Basically you want to make sure that what gets loaded in via JSON and what gets loaded in via YAML is the same. Once you have verified that in a test, you can just do all the edge case testing in one (say JSON). You don't need to duplicate them all for YAML as well.

File: src/bb2.py (hunk 1 above)
Comment (dtisza1) [reply src/bb2.py:10]: @nbragdon I was searching around to see what is typically used for packages and found this as one of the simpler ways being used. Not sure if it is the best.

However, I can see this not working, like in the case you mention. We should let the SDK user configure this path, in case this does not work for their setup. I'll add this shortly to the PR.

Also thinking of renaming this to "BASE_DIR".

File: src/bb2.py (hunk 2 above)
Comment (dtisza1) [reply src/bb2.py:23]: @nbragdon I will add this in. That was something I liked in the node SDK that you added. TY!

File: src/bb2.py (hunk 4 above)
Comment (dtisza1) [reply src/bb2.py:27]: @nbragdon Good idea to not save this, since it is only used for the base url. I'll remove this.  TY!

File: bluebutton-sample-config.json (hunk 5 above)
Comment (dtisza1) [reply bluebutton-sample-config.json:1]: @nbragdon YES! This is a good change. I'll remove these files and include them in the README.md. The SDK user can copy the example configs from there.  TY!

File: src/tests/test_configs/yaml/bluebutton-sample-config-missing-callback.yaml (hunk 6 above)
Comment (dtisza1) [reply src/tests/test_configs/yaml/bluebutton-sample-config-missing-callback.yaml:1]: @nbragdon I like this!  Will update the tests to do this.  TY!

File: src/bb2.py (hunk 3 above)
Comment (dtisza1) [reply src/bb2.py:66]: This is matching the new Node SDK behavior where there is an error, if the environment isn't explicitly set to PRODUCTION or SANDOBOX.

Comment (nbragdon) [reply src/bb2.py:66]: I guess I was thinking of the case where `environment` is just left off the config. Basically if a value is provided it should be checked against viable values. If it's left off we just default to sandbox. Probably not a big deal but is 1 less thing to get up and running quickly. Only when you transition to prod is it relevant.

File: src/bb2.py (hunk 1 above)
Comment (nbragdon) [reply src/bb2.py:10]: If there is no easy way to do it, I think this is fine. We'll just make sure to include it in the readme. Just wanted to make sure 👍

File: src/bb2.py [hunk 7]
```diff
@@ -2,13 +2,90 @@
...
+ROOT_DIR = os.path.abspath(os.curdir) + "/"
+DEFAULT_CONFIG_FILE_LOCATION = ROOT_DIR + "./.bluebutton-config.json"
+
+ENVIRONMENT_URLS = {
+    "SANDBOX": "https://sandbox.bluebutton.cms.gov",
+    "PRODUCTION": "https://api.bluebutton.cms.gov",
+}
 
 
 class Bb2:
```
Comment (nbragdon) [inline src/bb2.py:19]: Think we could also change this to `BlueButton`? Just to give consistency across all the projects

File: src/bb2.py (hunk 2 above)
Comment (dtisza1) [reply src/bb2.py]: @nbragdon Just added the option to pass in a dictionary.

File: src/bb2.py (hunk 4 above)
Comment (dtisza1) [reply src/bb2.py]: @nbragdon This is removed.

File: bluebutton-sample-config.json (hunk 5 above)
Comment (dtisza1) [reply bluebutton-sample-config.json:1]: @nbragdon Removed the sample config file.

File: src/tests/test_configs/yaml/bluebutton-sample-config-missing-callback.yaml (hunk 6 above)
Comment (dtisza1) [reply src/tests/test_configs/yaml/bluebutton-sample-config-missing-callback.yaml]: @nbragdon Updated the tests and removed  .yaml files for this. Left one YAML file for the JSON <-> YAML read_config() test.

This is much cleaner! TY!

File: src/bb2.py (hunk 7 above)
Comment (dtisza1) [reply src/bb2.py]: @nbragdon Keep forgetting to change this. Just changed it :-)

**Review by JFU-GIT — APPROVED** [review]
Looks good to me.

**Review by nbragdon — APPROVED** [review]
